src/hash_builder/mod.rs

```diff
@@ -372,7 +410,18 @@ impl HashBuilder {
             self.hash_masks[parent_index] |= TrieMask::from_nibble(current[parent_index]);
         }
 
-        let store_in_db_trie = !self.tree_masks[len].is_empty() || !self.hash_masks[len].is_empty();
+        // first set the state mask - must contain all child positions
+        let my_state_mask = self.groups[len];
+
+        // then determine tree mask - must be subset of state mask
+        let my_tree_mask = if self.stored_in_database {
+            // only set tree mask bits for positions that exist in state mask
+            my_state_mask & self.tree_masks[len]
```

Comment:
`self.tree_masks[len]` must always be a subset of `self.groups[len]`. it's currently validated below, why is this needed?
https://github.com/alloy-rs/trie/blob/cf984afb87802665c8e750f74dccc4b7c167c5f7/src/nodes/branch.rs#L285-L298